sklearn/utils/validation.py

```diff
@@ -31,28 +31,35 @@
 warnings.simplefilter('ignore', NonBLASDotWarning)
 
 
-def _assert_all_finite(X):
+def _assert_all_finite(X, allow_nan=False):
     """Like assert_all_finite, but only for ndarray."""
     if _get_config()['assume_finite']:
         return
     X = np.asanyarray(X)
     # First try an O(n) time, O(1) space solution for the common case that
     # everything is finite; fall back to O(n) space np.isfinite to prevent
     # false positives from overflow in sum method.
-    if (X.dtype.char in np.typecodes['AllFloat'] and not np.isfinite(X.sum())
-            and not np.isfinite(X).all()):
+    if (not allow_nan and X.dtype.char in np.typecodes['AllFloat']
+            and not np.isfinite(X.sum()) and not np.isfinite(X).all()):
         raise ValueError("Input contains NaN, infinity"
                          " or a value too large for %r." % X.dtype)
+    elif (allow_nan and X.dtype.char in np.typecodes['AllFloat']
+          and not np.isfinite(X[~np.isnan(X)].sum())
```

Comment:
np.isnan(X) already spends the memory that .sum() is trying to avoid. You do not need both the .sum() and the .all() conditions if you're using np.isnan(X). `if np.isnan().any()` is sufficient and, without chunking, optimal I think.

Reply:
I am not really sure what you mean. we would like to test for infinity only, therefore shall I do the following:

```python
if ... and not np.isfinite(x[~np.isnan(x)]).all():
```

Reply:
Sorry, I meant to say np.isinf

Reply:
It might be worth leaving in the O(1) memory case actually, as if all is finite, you can move on. Sorry for my mistakes above.

Reply:
Oh ok I see.